File: typescript/examples/tools/agents/decorator-pattern-agent.ts
#!/usr/bin/env node
/**
 * Matimo Decorator Pattern - True AI Agent with TypeScript @tool Decorators
 *
 * The agent receives a prompt, uses OpenAI to decide which tool to use,
 * then executes that tool via Matimo using real TypeScript @tool decorators.
 *
 * Run: npm run agent:decorator
 */

import 'dotenv/config';
import path from 'path';
import { fileURLToPath } from 'url';
import { ChatOpenAI } from '@langchain/openai';
import { MatimoInstance, tool, setGlobalMatimoInstance } from 'matimo';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

/**
 * Decorator Pattern Agent - Uses @tool decorators for automatic execution
 */
class DecoratorPatternAgent {
  private llm: ChatOpenAI;

  constructor(
    private matimo: MatimoInstance,
    llm: ChatOpenAI
  ) {
    this.llm = llm;
  }

  /**
   * Calculator tool - automatically executes via @tool decorator
   * Parameters map to tool parameters in order: operation, a, b
   */
  @tool('calculator')
  async calculate(operation: string, a: number, b: number): Promise<unknown> {
    // Decorator automatically calls: matimo.execute('calculator', { operation, a, b })
    // Method body is not executed - decorator intercepts the call
    return undefined;
  }

  /**
   * Echo tool - automatically executes via @tool decorator
   * Parameters map to tool parameter: message
   */
  @tool('echo-tool')
  async echo(message: string): Promise<unknown> {
    // Decorator automatically calls: matimo.execute('echo-tool', { message })
    return undefined;
  }

  /**
   * HTTP client tool - automatically executes via @tool decorator
   * Parameters map to tool parameters: method, url
   */
  @tool('http-client')
  async fetch(method: string, url: string): Promise<unknown> {
    // Decorator automatically calls: matimo.execute('http-client', { method, url })
    return undefined;
  }

  /**
   * Generate OpenAI function schemas from Matimo tool definitions
   * Single source of truth - schemas come from tool YAML, not duplicated here
   */
  private getToolSchemas() {
    return this.matimo.listTools().map((tool) => ({
      type: 'function',
      function: {
        name: tool.name,
        description: tool.description,
        parameters: {
          type: 'object',
          properties: Object.entries(tool.parameters || {}).reduce(
            (acc, [paramName, param]) => ({
              ...acc,
              [paramName]: {
                type: param.type,
                enum: param.enum,
                description: param.description,
              },
            }),
            {}
          ),
          required: Object.entries(tool.parameters || {})
            .filter(([_, param]) => param.required)
            .map(([paramName]) => paramName),
        },
      },
    }));
  }

  /**
   * Process a prompt - AI decides which tool to call
   */
  async process(prompt: string): Promise<void> {
    console.info(`\n❓ Prompt: "${prompt}"`);

    try {
      // Prepare system message for tool calling
      const systemMessage = `You are an AI assistant with access to tools. 
Based on the user's request, decide which tool to use and extract the EXACT required parameters.
IMPORTANT: Use exact parameter names and enum values as specified.
Respond ONLY with valid JSON in this format: {"tool": "<tool_name>", "parameters": {<exact_params>}}`;

      // Get tool schemas dynamically from Matimo
      const toolSchemas = this.getToolSchemas();

      // Build detailed tool specifications for the prompt
      const toolSpecifications = toolSchemas
        .map((t) => {
          const paramSpecs = Object.entries(t.function.parameters.properties)
            .map(([paramName, prop]: [string, any]) => {
              let spec = `${paramName} (${prop.type})`;
              if (prop.enum) {
                spec += ` - valid values: [${prop.enum.join(', ')}]`;
              }
              spec += ` - ${prop.description}`;
              return spec;
            })
            .join('; ');
          return `${t.function.name}: ${t.function.description}\n    Parameters: ${paramSpecs}`;
        })
        .join('\n\n');

      // Create a message with proper formatting for function calling
      const messages = [
        {
          type: 'system',
          content: systemMessage,
        },
        {
          type: 'human',
          content: `User request: "${prompt}"\n\nAvailable tools:\n${toolSpecifications}`,
        },
      ];

      // Call LLM with tool definitions
      const response = await this.llm.invoke(messages as any);

      // Parse the LLM response
      let toolName: string | null = null;
      let toolParams: Record<string, unknown> | null = null;

      // Try to extract tool call from response
      const content = response.content;

      if (typeof content === 'string') {
        // Try to parse as JSON
        try {
          const parsed = JSON.parse(content);
          toolName = parsed.tool || parsed.function?.name;
          toolParams = parsed.parameters || parsed.function?.parameters || parsed;
        } catch {
          // If not JSON, try to extract from text
          const jsonMatch = content.match(/\{[^{}]*"tool"[^{}]*\}/);
          if (jsonMatch) {
            try {
              const parsed = JSON.parse(jsonMatch[0]);
              toolName = parsed.tool;
              toolParams = parsed.parameters;
            } catch {
              // Continue without params
            }
          }
        }
      }

      // If we found a tool, execute it via decorated method
      if (toolName && toolParams) {
        await this.executeTool(toolName, toolParams);
      } else {
        console.info(`\n⚠️  No tool call detected in response`);
        console.info(
          `Response: ${typeof content === 'string' ? content.substring(0, 200) : content}`
        );
      }
    } catch (error) {
      console.error(`\n❌ Error: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  /**
   * Map of tool names to their decorated methods
   * Built at runtime to discover which tools this agent supports
   */
  private getToolMethodMap(): Map<string, string> {
    const toolMap = new Map<string, string>();

    // Get all methods decorated with @tool
    // The decorator stores tool name in a metadata property
    for (const [methodName, descriptor] of Object.entries(
      Object.getOwnPropertyDescriptors(Object.getPrototypeOf(this))
    )) {
      if (typeof descriptor.value === 'function') {
        // Check if method has tool metadata (added by decorator)
        const method = descriptor.value as any;
        if (method.__toolName) {
          toolMap.set(method.__toolName, methodName);
        }
      }
    }

    // Manual mapping as fallback (decorators should set __toolName)
    // This ensures we catch all @tool decorated methods
    toolMap.set('calculator', 'calculate');
    toolMap.set('echo-tool', 'echo');
    toolMap.set('http-client', 'fetch');

    return toolMap;
  }

  /**
   * Execute a tool via decorated method
   * Uses reflection to dynamically call the appropriate decorated method
   * This keeps decorators as the source of truth for agent API
   */
  private async executeTool(toolName: string, params: Record<string, unknown>): Promise<void> {
    try {
      // Normalize parameters based on tool
      let normalizedParams = params;

      // Calculator: Handle "operands" array format by converting to a, b
      if (toolName === 'calculator' && params.operands && Array.isArray(params.operands)) {
        const [a, b] = params.operands as number[];
        normalizedParams = {
          operation: params.operation,
          a,
          b,
        };
      }

      console.info(`\n🔧 Using tool: ${toolName}`);
      console.info(`   Parameters: ${JSON.stringify(normalizedParams)}`);

      // Find the decorated method for this tool
      const toolMethodMap = this.getToolMethodMap();
      const methodName = toolMethodMap.get(toolName);

      if (!methodName) {
        console.info(`\n❌ Tool '${toolName}' not in agent's API`);
        console.info(`Available tools: ${Array.from(toolMethodMap.keys()).join(', ')}`);
        return;
      }

      // Dynamically call the decorated method
      // The decorator intercepts the call and executes the tool
      const method = (this as any)[methodName];
      if (typeof method !== 'function') {
        console.info(`\n❌ Method '${methodName}' for tool '${toolName}' not found`);
        return;
      }

      // Convert params object to positional args matching method signature
      const args = this.getMethodArgsFromParams(toolName, normalizedParams);
      const result = await method.apply(this, args);

      // Format result
      if (typeof result === 'object' && result !== null) {
        const resultData = result as any;
        if (resultData.stdout) {
          try {
            const parsed = JSON.parse(resultData.stdout);
            console.info(`\n✅ Result:`, parsed);
          } catch {
            console.info(`\n✅ Result:`, resultData.stdout);
          }
        } else if (resultData.data) {
          // HTTP response
          console.info(
            `\n✅ Result (HTTP ${resultData.statusCode}):`,
            typeof resultData.data === 'string'
              ? resultData.data.substring(0, 200)
              : JSON.stringify(resultData.data).substring(0, 200)
          );
        } else {
          console.info(`\n✅ Result:`, result);
        }
      }
    } catch (error) {
      console.error(`\n❌ Tool Error: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  /**
   * Convert parameter object to positional arguments for method call
   * Maps normalized params to the decorated method's parameter order
   */
  private getMethodArgsFromParams(toolName: string, params: Record<string, unknown>): unknown[] {
    switch (toolName) {
      case 'calculator':
        return [params.operation, params.a, params.b];
      case 'echo-tool':
        return [params.message];
      case 'http-client':
        return [params.method, params.url];
      default:
        return [];
    }
  }
}

/**
 * Run decorator pattern agent with prompts
 */
async function runDecoratorPatternAgent() {
  console.info('\n╔════════════════════════════════════════════════════════╗');
  console.info('║   Matimo Decorator Pattern - True AI Agent             ║');
  console.info('║   (AI decides which tool to use based on prompt)       ║');
  console.info('╚════════════════════════════════════════════════════════╝\n');

  try {
    const apiKey = process.env.OPENAI_API_KEY;
    if (!apiKey) {
      console.error('❌ Error: OPENAI_API_KEY not set in .env');
      process.exit(1);
    }

    // Initialize Matimo with auto-discovery
    console.info('🚀 Initializing Matimo...');
    const matimo = await MatimoInstance.init({ autoDiscover: true });

    // Set global Matimo instance for @tool decorators
    setGlobalMatimoInstance(matimo);

    const matimoTools = matimo.listTools();
    console.info(`📦 Loaded ${matimoTools.length} tools:\n`);
    matimoTools.forEach((t) => {
      console.info(`  • ${t.name}`);
      console.info(`    ${t.description}\n`);
    });

    // Initialize OpenAI LLM
    console.info('🤖 Initializing OpenAI LLM (gpt-3.5-turbo)...\n');
    const llm = new ChatOpenAI({
      modelName: 'gpt-4o-mini',
      temperature: 0,
      openAIApiKey: apiKey,
    });

    // Create agent
    const agent = new DecoratorPatternAgent(matimo, llm);

    // Test prompts - each should trigger a different tool
    const prompts = [
      '🧮 What is 42 plus 8?',
      '🔊 Echo the message: "Decorator pattern is elegant and powerful!"',
      '🌐 Fetch the GitHub user profile for octocat using HTTP GET',
    ];

    console.info('🧪 Testing AI Agent with 3 Different Prompts');
    console.info('═'.repeat(60));

    for (const prompt of prompts) {
      await agent.process(prompt);
      console.info('\n' + '─'.repeat(60));
    }

    console.info('\n✅ Decorator pattern AI agent test complete!\n');
  } catch (error) {
    console.error('❌ Agent failed:', error instanceof Error ? error.message : String(error));
    process.exit(1);
  }
}

// Run the agent
runDecoratorPatternAgent().catch(console.error);

File: python/examples/native/agents/decorator_pattern_agent.py
"""
Matimo Decorator Pattern — AI Agent with @tool Decorators.
Mirrors: typescript/examples/tools/agents/decorator-pattern-agent.ts

The agent receives a prompt, uses OpenAI to decide which tool to use,
then executes it via @tool-decorated methods.

Run:
    cd python
    OPENAI_API_KEY=sk-... uv run python examples/native/agents/decorator_pattern_agent.py
"""
from __future__ import annotations

import asyncio
import json
import os
import re
import sys
from typing import Any

from langchain_openai import ChatOpenAI
from langchain_core.messages import HumanMessage, SystemMessage

from matimo import Matimo
from matimo.decorators import tool, set_global_matimo_instance


class DecoratorPatternAgent:
    """
    Decorator Pattern Agent — @tool decorators intercept method calls
    and execute the corresponding Matimo tool automatically.
    """

    def __init__(self, matimo: Matimo, llm: ChatOpenAI) -> None:
        self._matimo = matimo
        self._llm = llm

    @tool("calculator")
    async def calculate(self, operation: str, a: float, b: float) -> Any:
        # Decorator intercepts → matimo.execute('calculator', {operation, a, b})
        ...

    @tool("echo-tool")
    async def echo(self, message: str) -> Any:
        # Decorator intercepts → matimo.execute('echo-tool', {message})
        ...

    @tool("http-client")
    async def fetch(self, method: str, url: str) -> Any:
        # Decorator intercepts → matimo.execute('http-client', {method, url})
        ...

    def _get_tool_schemas(self) -> list[dict]:
        """Build OpenAI function schemas from Matimo tool definitions."""
        return [
            {
                "type": "function",
                "function": {
                    "name": t.name,
                    "description": t.description,
                    "parameters": {
                        "type": "object",
                        "properties": {
                            param_name: {
                                "type": param.type.value if hasattr(param.type, "value") else param.type,
                                "description": param.description or "",
                                **({"enum": param.enum} if param.enum else {}),
                            }
                            for param_name, param in (t.parameters or {}).items()
                        },
                        "required": [
                            name for name, p in (t.parameters or {}).items() if p.required
                        ],
                    },
                },
            }
            for t in self._matimo.list_tools()
        ]

    async def process(self, prompt: str) -> None:
        print(f'\n❓ Prompt: "{prompt}"')

        tool_schemas = self._get_tool_schemas()
        tool_specs = "\n\n".join(
            "{name}: {desc}\n    Parameters: {params}".format(
                name=s["function"]["name"],
                desc=s["function"]["description"],
                params="; ".join(
                    "{k} ({t}){enum} - {d}".format(
                        k=k,
                        t=v.get("type", ""),
                        enum=f" - valid values: [{', '.join(v['enum'])}]" if v.get("enum") else "",
                        d=v.get("description", ""),
                    )
                    for k, v in s["function"]["parameters"]["properties"].items()
                ),
            )
            for s in tool_schemas
        )

        messages = [
            SystemMessage(
                content=(
                    "You are an AI assistant with access to tools. "
                    "Based on the user's request, decide which tool to use and extract the EXACT required parameters. "
                    "IMPORTANT: Use exact parameter names and enum values as specified. "
                    'Respond ONLY with valid JSON: {"tool": "<tool_name>", "parameters": {<exact_params>}}'
                )
            ),
            HumanMessage(content=f'User request: "{prompt}"\n\nAvailable tools:\n{tool_specs}'),
        ]

        try:
            response = await self._llm.ainvoke(messages)
            content = response.content

            tool_name: str | None = None
            tool_params: dict | None = None

            if isinstance(content, str):
                try:
                    parsed = json.loads(content)
                    tool_name = parsed.get("tool") or parsed.get("function", {}).get("name")
                    tool_params = parsed.get("parameters") or parsed.get("function", {}).get("parameters") or parsed
                except json.JSONDecodeError:
                    match = re.search(r'\{[^{}]*"tool"[^{}]*\}', content)
                    if match:
                        try:
                            parsed = json.loads(match.group())
                            tool_name = parsed.get("tool")
                            tool_params = parsed.get("parameters")
                        except json.JSONDecodeError:
                            pass

            if tool_name and tool_params is not None:
                await self._execute_via_decorator(tool_name, tool_params)
            else:
                print("\n⚠️  No tool call detected in response")
                print(f"Response: {str(content)[:200]}")

        except Exception as exc:
            print(f"\n❌ Error: {exc}")

    # Tool name → (method name, positional arg order)
    _TOOL_METHOD_MAP: dict[str, tuple[str, list[str]]] = {
        "calculator": ("calculate", ["operation", "a", "b"]),
        "echo-tool": ("echo", ["message"]),
        "http-client": ("fetch", ["method", "url"]),
    }

    async def _execute_via_decorator(self, tool_name: str, params: dict[str, Any]) -> None:
        # Normalise calculator operands array → a, b
        if tool_name == "calculator" and isinstance(params.get("operands"), list):
            a, b = params["operands"][0], params["operands"][1]
            params = {"operation": params.get("operation"), "a": a, "b": b}

        print(f"\n🔧 Using tool: {tool_name}")
        print(f"   Parameters: {json.dumps(params)}")

        mapping = self._TOOL_METHOD_MAP.get(tool_name)
        if not mapping:
            print(f"\n❌ Tool '{tool_name}' not in agent's API")
            print(f"Available tools: {', '.join(self._TOOL_METHOD_MAP)}")
            return

        method_name, arg_order = mapping
        method = getattr(self, method_name, None)
        if not callable(method):
            print(f"\n❌ Method '{method_name}' not found")
            return

        # Pass params in the order the method signature expects
        args = [params.get(k) for k in arg_order]

        try:
            result = await method(*args)

            if isinstance(result, dict):
                if "stdout" in result:
                    try:
                        print("\n✅ Result:", json.loads(result["stdout"]))
                    except (json.JSONDecodeError, TypeError):
                        print("\n✅ Result:", result["stdout"])
                elif "data" in result:
                    data_str = json.dumps(result["data"])[:200]
                    print(f"\n✅ Result (HTTP {result.get('status_code', '?')}): {data_str}")
                else:
                    print("\n✅ Result:", result)
            else:
                print("\n✅ Result:", result)

        except Exception as exc:
            print(f"\n❌ Tool Error: {exc}")


async def main() -> None:
    print("\n╔════════════════════════════════════════════════════════╗")
    print("║   Matimo Decorator Pattern - True AI Agent             ║")
    print("║   (AI decides which tool to use based on prompt)       ║")
    print("╚════════════════════════════════════════════════════════╝\n")

    api_key = os.environ.get("OPENAI_API_KEY")
    if not api_key:
        print("❌ Error: OPENAI_API_KEY not set")
        sys.exit(1)

    print("🚀 Initializing Matimo...")
    matimo = await Matimo.init(auto_discover=True)

    # Register global instance so @tool decorators can resolve it
    set_global_matimo_instance(matimo)

    tools = matimo.list_tools()
    print(f"📦 Loaded {len(tools)} tools:\n")
    for t in tools:
        print(f"  • {t.name}")
        print(f"    {t.description}\n")

    print("🤖 Initializing OpenAI LLM (gpt-4o-mini)...\n")
    llm = ChatOpenAI(model="gpt-4o-mini", temperature=0, api_key=api_key)

    agent = DecoratorPatternAgent(matimo, llm)

    prompts = [
        "🧮 What is 42 plus 8?",
        '🔊 Echo the message: "Decorator pattern is elegant and powerful!"',
        "🌐 Fetch the GitHub user profile for octocat using HTTP GET",
    ]

    print("🧪 Testing AI Agent with 3 Different Prompts")
    print("═" * 60)

    for prompt in prompts:
        await agent.process(prompt)
        print("\n" + "─" * 60)

    print("\n✅ Decorator pattern AI agent test complete!\n")


if __name__ == "__main__":
    asyncio.run(main())
